include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -20,15 +20,10 @@
 #include <type_traits>
 #include <tuple>
 
-#if _ONEDPL___cplusplus >= 202002L && __has_include(<bit>)
-#    include <bit>
-#else
-#    include <cstring> // memcpy
-#endif
-
 #include "../../iterator_impl.h"
 
 #include "sycl_defs.h"
+#include "execution_sycl_defs.h"
```

Comment:
missing includes: `<algorithm>` `<cstddef>`, `<string>`, `<cstdint>`

Reply:
Well, except for `<algorithm>` for `std::min`, all the rest of standard types are used in consistence with the SYCL spec - so it should be safe to assume that the necessary headers are included through the SYCL headers. Is it necessary to duplicate? In the lines 77-80, `std::vector` is used; should we also explicitly include `<vector>`?

I can change most or even all these std types to `auto` ("almost always auto", yeah) to formally have no need to include extra headers. Do you think it would be better?

Reply:
I think so. The potential problem could be after we refactor the code we might end up depending on the include order. I hope that sycl includes everything but nobody prevents them to refactor their code (unless headers are explicitly listed in the SYCL spec). I prefer to do headers self-contained because it's just safer.

Reply:
You understand that technically `auto` changes nothing with regard to headers being self-contained, right? The functionality does not change, no new headers are included, and we still rely on SYCL implementation to do the proper job (while "nobody prevents them to refactor their code"). The only thing is that we **formally** do not use standard types. meanwhile losing the information on the return types of the routines.

Reply:
Probably with auto you are right. If they don't include proper file, the code with auto doesn't work at all. I also understand what you mean: when SYCL returns some type it should have it's definition, on other words, include a proper header. Otherwise, it cannot be used. (Let us not to go in forward declaration details).

What I am struggling with is how to review the code? we have explicit usage of types in that. What should we do? Look it that's what we are using and distinguish it from types that other framework returns?

Reply:
Yes, I understand this concern. I think a careful reviewer can recognize where the use of certain types (standard or not) is associated with some API (and so implicitly covered by an already included header) versus where the types are used "on their own" and so a header is needed. But definitely it's extra effort, and some potential issues might be missed.

My viewpoint is, though, that a probability of something going wrong some time in the future is always non-zero, and trying to make the code resilient to future changes and possible mistakes is not always worth it. We need to make sure the code is free of actual errors, while robustness against potential future errors is a second-order concern at best.

In this very case, absence of standard headers is not an error because the standard types are used in the context of SYCL API.